File: packages/py/src/airlock/application/ports/event_bus.py
"""Event-bus ports. The publish and subscribe sides are separate (Interface
Segregation): the agent core depends only on EventPublisher; driving adapters
(the runner, approvers) depend on EventSubscriber."""

from collections.abc import Awaitable, Callable
from typing import Protocol

from pydantic import BaseModel

from airlock.domain.events import EventTopic

EventHandler = Callable[[object], Awaitable[None] | None]
"""Receives a JSON-compatible payload (a dict) and handles it."""


class EventPublisher(Protocol):
    async def publish(self, topic: EventTopic, event: BaseModel) -> None: ...


class EventSubscriber(Protocol):
    async def subscribe(self, topic: EventTopic, handler: EventHandler) -> None: ...

File: packages/ts/src/application/ports/event-bus.ts
import type { EventTopic } from "../../domain/events";

export type EventHandler<TEvent> = (event: TEvent) => Promise<void> | void;

/**
 * Publish side of the event bus. The agent core depends only on this
 * (Interface Segregation): it emits approval requests and lifecycle events.
 */
export interface EventPublisher {
  publish<TEvent>(topic: EventTopic, event: TEvent): Promise<void>;
}

/**
 * Subscribe side of the event bus. Driving adapters (the runner, approvers)
 * depend on this; the core does not.
 */
export interface EventSubscriber {
  subscribe<TEvent>(topic: EventTopic, handler: EventHandler<TEvent>): Promise<void>;
}
